\documentclass[11pt,twoside,a4paper]{article}
\usepackage{pslatex,palatino,avant,graphicx}
\usepackage[usenames,dvipsnames]{color}
\usepackage[margin=2cm]{geometry}
\usepackage{url}
\usepackage[square,sort]{natbib}

\usepackage{listings}
\lstset{ %
  backgroundcolor=\color{white},   % choose the background color; you must add \usepackage{color} or \usepackage{xcolor}
  basicstyle=\footnotesize,        % the size of the fonts that are used for the code
  breakatwhitespace=false,         % sets if automatic breaks should only happen at whitespace
  breaklines=true,                 % sets automatic line breaking
  captionpos=b,                    % sets the caption-position to bottom
  commentstyle=\color{LimeGreen},  % comment style
  % deletekeywords={...},            % if you want to delete keywords from the given language
  escapeinside={\%*}{*)},          % if you want to add LaTeX within your code
  extendedchars=true,              % lets you use non-ASCII characters; for 8-bits encodings only, does not work with UTF-8
  frame=single,                    % adds a frame around the code
  keepspaces=true,                 % keeps spaces in text, useful for keeping indentation of code (possibly needs columns=flexible)
  keywordstyle=\color{BlueGreen},  % keyword style
  % language=Java,                   % the language of the code
  % numbers=left,                    % where to put the line-numbers; possible values are (none, left, right)
  % numbersep=5pt,                   % how far the line-numbers are from the code
  % numberstyle=\tiny\color{Gray},   % the style that is used for the line-numbers
  rulecolor=\color{black},         % if not set, the frame-color may be changed on line-breaks within not-black text (e.g. comments (green here))
  showspaces=false,                % show spaces everywhere adding particular underscores; it overrides 'showstringspaces'
  showstringspaces=false,          % underline spaces within strings only
  showtabs=false,                  % show tabs within strings adding particular underscores
  stepnumber=2,                    % the step between two line-numbers. If it's 1, each line will be numbered
  stringstyle=\color{RubineRed},   % string literal style
  tabsize=2,                       % sets default tabsize to 2 spaces
  title=\lstname,                  % show the filename of files included with \lstinputlisting; also try caption instead of title
  aboveskip=1em,
  belowcaptionskip=0em,
  belowskip=0em
}

\usepackage{hyperref}
\hypersetup{
    colorlinks,
    citecolor=Violet,
    filecolor=black,
    linkcolor=MidnightBlue,
    urlcolor=MidnightBlue
}

\setlength\parindent{0pt}

\begin{document}

\providecommand{\versionnumber}{0.1.1}
\title{PartitionTest 1.0 Manual v\versionnumber}
\author{Diego Darriba, David Posada}
\date{November 6, 2015}
\maketitle

\setcounter{tocdepth}{2}
\tableofcontents

\section{Overview}

PartitionTest is a tool to carry out statistical selection of best-fit partitioning schemes and models of nucleotide substitution / amino acid replacement.

\subsection{Download}

The main project webpage is located at GitHub: \url{http://www.github.com/ddarriba/partitiontest}.

\subsection{Installation}

PartitionTest depends on the Phylogenetic Likelihood Library (PLL) 1.0.11+.
You can download PLL library at \url{http://www.libpll.org}.

First, you need to install PLL following the instructions included in the package. Typically:

\begin{verbatim}
$ wget http://libpll.org/Downloads/libpll-1.0.11.tar.gz
$ tar zvxf libpll-1.0.11.tar.gz
$ cd libpll-1.0.11
$ ./configure && make && make install
\end{verbatim}

Next, download and unpackage PartitionTest and follow the instructions in the INSTALL file:

\begin{verbatim}
$ tar zvxf partitiontest-xx.yy.zz.tar.gz
$ cd partitiontest-xx.yy.zz
$ ./configure && make && make install
\end{verbatim}

Finally, you can check the installation by running an example data set:

\begin{verbatim}
$ partest -c example-data/dna/example1.conf
\end{verbatim}

Obtain command-line help by using the {\it help} argument or the {\it man} page:

\begin{verbatim}
$ partest --help
$ man partest
\end{verbatim}

%\subsection{Citation}

%-

\subsection{Disclaimer}

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 3 of the License, or (at your option) any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. The jModelTest distribution includes Phyml executables.

These programs are protected by their own license and conditions, and using jModelTest implies agreeing with those conditions as well. 

\clearpage
\section{Usage}

Typical usage of PartitionTest involves the creation of a configuration file.
Although not mandatory, because PartitionTest can do a model selection on single-partition
data sets, the configuration file is necessary for defining the initial datablocks.
All parameters specified by command line will take preference over the ones specified in
the configuration file.

\subsection{Configuration file}

Running PartitionTest with the argument ``- - config-help'' prints out the format of the
configuration file. The argument ``- - config-template'' prints out a template for the
configuration file that you can redirect to a file and fill it up with your own options.

The format is the following:

\begin{verbatim}
; This is a comment

; Start of input data parameters
[input]
msa           = MSA_FILENAME
tree          = TREE_FILENAME (default:maximum parsimony)
topo	         = {mp|ml|fixed}
datatype      = {nt|aa} (default: nt)
keep-branches = {true|false} (default: false)

; Start of searching options
[search]
algorithm  = {greedy|hcluster|random|k1|kn|auto} (default: auto)
replicates = # (default: 1)

; Start of candidate models description
[models]
include = {all | gtr | [LIST]} (default:all)

; Start of partitions
[partitions]
DATABLOCK1 = INI1-END1
DATABLOCK2 = INI2-END2
...
DATABLOCKn = INIn-ENDn

; Start of schemes
[schemes]
S1 = (DATABLOCK1,DATABLOCK2)(DATABLOCK3)...
S2 = (DATABLOCK1,DATABLOCK2,DATABLOCK3)...
...
Sn = (DATABLOCK1)(DATABLOCK2,DATABLOCK3,...)

; Start of output section
[output]
path=OUTPUT_BASE_URL (default:partitiontest_[MSA_FILENAME]
\end{verbatim}

\subsection{Command line arguments}

Mandatory arguments to long options are mandatory for short options too

\begin{flushleft}
\renewcommand{\arraystretch}{1.5}
\small {
\begin{tabular}{@{}llp{.7\linewidth}@{}}
-c & - - config-file {\it filename}
   &   Sets the input configuration file. Run with - - config-help for more information \\

 &- - config-help 
 &      Shows help about configuration files \\

 & - - config-template 
 &      Generates a configuration file template \\

-d& - - data-type {\it datatype} 
  &     Sets the type of the input data. DATA\_TYPE: nt (nucleotide), aa (amino-acid) \\

  &- - disable-ckp 
  &     Disables the checkpointing \\

  &- - disable-output 
  &     Disables any file-based output. This option also disables the chackpointing\\

-F & - - empirical-frequencies
   &    Includes models with empirical frequencies (+F) \\

-g& - - pergene-bl
  &     Estimate per-gene branch-lengths \\

&- - force-override
&Existent output files will be overwritten \\

-h& - - help-
  &     Displays a help message \\

-i& - - input-file {\it filename}
  &     Sets the input alignment file ({\bf required}) \\

-k& - - keep-branches
  &     Keep branch lengths from the initial topology. This argument has no effect for initial topology different than fixed \\

-N& - - non-stop
  &     Algorithms do not stop if no improvement found at one step, for avoiding local maxima \\

-O& - - optimize-mode {\it mode}
  &     Sets  the  model optimization for the best-fit partition. {\it mode}: findmodel (find the best-fit model for each partition), gtr (use only GTR model for each partition (nucleic data)
or AUTO for protein data) \\

-p& - - num-procs {\it \#}
  &     Number of threads for model evaluation (DEFAULT: 1) \\

-r& - - replicates {\it \#}
  &     Sets the number of replicates on Hierarchical Clustering and Random search modes \\

-s& - - selection-criterion {\it sc}
  &     Sets the criterion for model selection. {\it sc}: bic (Bayesian Information Criterion) (DEFAULT), aic (Akaike Information Criterion), aicc (Corrected Akaike Information  Criterion),  dt
(Decision Theory). Sample size for bic, aicc and dt criteria is the alignment length \\

-S& - - search {\it algorithm}
  &     Sets  the  search  algorithm  {\it algorithm}:  k1  (evaluate K=1 only), kn (evaluate K=N only), greedy (greedy search algorithm), greedyext (extended greedy search algorithm), hcluster
(hierarchical clustering algorithm), random (multiple step random sampling), auto (auto-select algorithm (DEFAULT)), exhaustive (exhaustive search) \\

-t& - - topology {\it type}
  &     Sets the starting topology for optimization. {\it type}: mp (creates a maximum parsimony topology for each model optimization (DEFAULT)), ml (creates a maximum likelihood topology
for each model optimization), fixed (uses a fixed ML topology for every model optimization), user (uses a user-defined topology. Requires the "-u" argument). However, if "-u" argument is
used this option is automatically set \\

-T& - - get-final-tree
  &     Conduct final ML tree optimization \\

-u& - - user-tree {\it filename}
  &     Sets a user-defined topology. This option ignores all starting topologies different from "user-defined". The tree must be in Newick format \\

-v& - - version
  &     Output version information and exit \\
  
-w& - - weights {\it w\_r,w\_f,w\_a}
  &     Sets the weights for computing partition distances: 
  {\it w\_r} = substitution rates matrix, 
  {\it w\_f} = base frequencies,
  {\it w\_a} = $\Gamma$ shape parameter ($\alpha$). \\
\end{tabular}
}
\end{flushleft}

\vspace{1em}
\subsection{Exit status}
\begin{itemize}
\item[0] OK
\item[64] Usage error
\item[65] Error with data
\item[69] Unavailable feature
\item[70] Internal software error (probably a bug)
\item[74] I/O error
\item[78] Error in configuration
\end{itemize}

%\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
\bibliographystyle{plain}
\bibliography{biblio}

\end{document}
